src/operator/tensor/elemwise_unary_op.h

```diff
@@ -257,39 +303,14 @@ void CastStorageDnsRspImpl(mshadow::Stream<xpu> *s, const TBlob& dns, NDArray* r
   CHECK_EQ(rsp->storage_type(), kRowSparseStorage);
   CHECK_EQ(dns.shape_, rsp->shape());
 
-  rsp->CheckAndAllocAuxData(rowsparse::kIdx, mshadow::Shape1(dns.shape_[0]));
   MSHADOW_TYPE_SWITCH(dns.type_flag_, DType, {  // data type
-    NDARRAY_IDX_TYPE_SWITCH(rsp->aux_type(rowsparse::kIdx), RType, {  // row idx type
-      RType* row_idx = rsp->aux_data(rowsparse::kIdx).dptr<RType>();
+    MSHADOW_INT_TYPE_SWITCH(rsp->aux_type(rowsparse::kIdx), RType, {  // row idx type
       const index_t num_rows = dns.shape_[0];
       const index_t num_cols = dns.shape_[1];
-      // Fill input_data.shape_[0] into row_idx array
-      mxnet_op::Kernel<FillRspRowIdx, xpu>::Launch(s, num_rows, row_idx, dns.dptr<DType>(),
-          num_rows, num_cols);
-
-      // single thread scanning row_idx array to find out number of non-zero rows
-      index_t nnr = 0;  // number of non-zero rows
-      for (index_t i = 0; i < num_rows; ++i) {
-        if (row_idx[i] < static_cast<RType>(num_rows)) ++nnr;
-      }
-      if (0 == nnr) {
-        rsp->SetAuxShape(rowsparse::kIdx, TShape(mshadow::Shape1(0)));
-        return;  // zero matrix
-      }
-      rsp->CheckAndAllocData(mshadow::Shape2(nnr, num_cols));
-      // TODO(junwu): single thread for compressing row_idx and copying data
-      // from dns to rsp, might be a bottleneck.
-      auto in_tensor = dns.FlatTo2D<xpu, DType>(s);
-      auto out_tensor = rsp->data().FlatTo2D<xpu, DType>(s);
-      int last_nnr_id = -1;  // last non-zero row id
-      for (index_t i = 0; i < num_rows; ++i) {
-        if (row_idx[i] < static_cast<RType>(num_rows)) {  // non-zero row found
-          row_idx[++last_nnr_id] = row_idx[i];
-          mshadow::Copy(out_tensor[last_nnr_id], in_tensor[i], s);
-        }
-      }
```

Comment:
We should still keep this function somewhere. if we need to make a row_sparse NDArray more compact, we can make a copy of the idx and data in a more memory efficient way.